lib/http2.c

```diff
@@ -2037,124 +2045,128 @@ static ssize_t cf_h2_send(struct Curl_cfilter *cf, struct Curl_easy *data,
       nwritten = http2_handle_stream_close(cf, data, stream, err);
       goto out;
     }
-    /* If stream_id != -1, we have dispatched request HEADERS, and now
-       are going to send or sending request body in DATA frame */
-    nwritten = Curl_bufq_write(&stream->sendbuf, buf, len, err);
-    if(nwritten < 0) {
-      if(*err != CURLE_AGAIN)
+    else if(stream->upload_blocked_len) {
+      /* the data in `buf` has alread been submitted or added to the
+       * buffers, but have been EAGAINed on the last invocation. */
+      DEBUGASSERT(len >= stream->upload_blocked_len);
+      if(len < stream->upload_blocked_len) {
+        /* Did we get called again with a smaller `len`? This should not
+         * happend. We are not prepared to handle that. */
+        failf(data, "HTTP/2 send again with decreased length");
```

Comment:
Hi @katrinleinweber, thanks for linking the details. Clearly, something is going wrong here. (btw. `cf` is our internal shortener of `connection filter`). From the lines I see that you are operating on 8.2.0.

I have to investigate what may be causing this. The check is there to rather fail than proceed with an upload that looks gone wrong. But, as usual, there seem to be wrong assumptions involved.

Reply:
Thanks for looking into this, @icing!

> you are operating on 8.2.0

No, I just wanted to comment at the source. [8.2.1 seems to have moved the same code downwards a bit](https://github.com/curl/curl/blob/curl-8_2_1/lib/http2.c#L2077-L2084).

Reply:
@katrinleinweber I tried in many ways to trigger this on my machine, but no success. I now suspect that this only happens on 32bit systems. Do you know of any reports on 64bit archs?

Reply:
I recently encountered this error with a `git push` to github.com. I'm not behind any proxies, and had the same result when I used a VPN to connect to an alternative github.com endpoint in another country.

I opened a support ticket with GitHub and they referenced this discussion and suggested I share my experience here. I am using 64 bit git.exe on Windows, version 2.42.0.windows.1.

I have captured a trace log with:

```
export GIT_TRACE_PACKET=1
export GIT_TRACE=1
export GIT_CURL_VERBOSE=1
```

I am no longer able to replicate the issue locally (temporarily turning off git compression tweaked things enough that the problem went away), but in case the logs are useful, I have attached them here ([git_connect.log.txt](https://github.com/curl/curl/files/12866733/git_connect.log.txt)). Short log:

```
mcdurdin@THARK MINGW64 /c/Projects/keyman/app (chore/core/9722-rename-km_kbp_-to_km_core_)
$ git push -u origin chore/core/9722-rename-km_kbp_-to_km_core_ 
Enumerating objects: 399, done.
Counting objects: 100% (399/399), done.
Delta compression using up to 8 threads
Compressing objects: 100% (264/264), done.
Writing objects: 100% (267/267), 63.66 KiB | 1.27 MiB/s, done.
Total 267 (delta 230), reused 0 (delta 0), pack-reused 0
error: RPC failed; curl 16 HTTP/2 send again with decreased length
send-pack: unexpected disconnect while reading sideband packet
fatal: the remote end hung up unexpectedly
Everything up-to-date
```

In the log, I find the `Send header` git trace immediately before the failure curious -- why would git be sending a zero-byte header?

```
14:54:19.988680 http.c:792              => Send header, 0000000000 bytes (0x00000000)
```

Reply:
> I recently encountered this error with a `git push` to github.com. I'm not behind any proxies, and had the same result when I used a VPN to connect to an alternative github.com endpoint in another country.
> 
> I opened a support ticket with GitHub and they referenced this discussion and suggested I share my experience here. I am using 64 bit git.exe on Windows, version 2.42.0.windows.1.

The issue was fixed in curl 8.3.0 released in September. If you are using the latest Git for Windows it was released in August so unless it has an auto-update of components then it is not using curl 8.3.0.